Fix inventory cache interface

s-hertel opened the pull request:
##### SUMMARY
There are a number of problems with the current cache interface, the primary ones being that the developer interface is terrible and it only handles file-backed cache plugins.

Some of the interface problems are that the cache was difficult to set up, get() was expected to raise a KeyError, and the cache had to be set a particular way to be reflected in the backing plugin.

This allows self._cache to be used as a dictionary - e.g. if you want a KeyError, use ``__getattr__`` instead of ``get``. It also updates the backing cache plugin with the value of self._cache if it has been modified after parse is called so the developer doesn't need to worry about the discrepancy between the cache object being exposed (a dictionary) and the permanent cache store (CacheModule).

Added tests and developer documentation.

##### ISSUE TYPE
- Bugfix Pull Request
- Feature Pull Request

s-hertel committed:
Replace InventoryFileCacheModule with a better developer-interface

Use new interface for inventory plugins with backwards compatibility

Auto-update the backing cache-plugin if the cache has changed after parsing the inventory plugin

s-hertel committed:
Update the inventory_cache documentation

s-hertel committed:
Update CacheModules to use config and add a deprecation warning if they are being imported directly rather than using cache_loader

s-hertel committed:
Fix foreman inventory caching

s-hertel committed:
Add tests

s-hertel committed:
Add developer docs

s-hertel committed:
Replace FactCache - undo or add backwards compatibility and deprecation period before merging

s-hertel commented on lib/ansible/vars/manager.py:
```diff
@@ -39,7 +39,9 @@
 from ansible.module_utils.six import iteritems, text_type, string_types
 from ansible.plugins.loader import lookup_loader, vars_loader
-from ansible.vars.fact_cache import FactCache
+# I'll undo because it might break 3rd party plugins (or could use a toggle + deprecation period?), but this works
```
DON'T MERGE - just want to run tests on shippable before I remove this. I'm also interested in opinions about deprecating FactCache. :-)

s-hertel committed:
documentation - fix --flush-cache and refresh_inventory

s-hertel committed:
appease shippable

s-hertel committed:
Undo FactCache

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/101019/60/tests

s-hertel committed:
Fix backwards compat for places using cache_loader without class_only

s-hertel committed:
Use class_only for fact_cache so the definitions are loaded before invoking the object. Now set_options/config actually works.

s-hertel commented on lib/ansible/vars/fact_cache.py:
```diff
@@ -21,7 +21,7 @@ class FactCache(MutableMapping):
     def __init__(self, *args, **kwargs):
 
-        self._plugin = cache_loader.get(C.CACHE_PLUGIN)
+        self._plugin = cache_loader.get(C.CACHE_PLUGIN, class_only=True)()
```
I'm not sure how this was intended to be used and therefore what the right fix is. The alternative solution for this (and [this](https://github.com/ansible/ansible/pull/50446/files#diff-a8385538c61f1318541f0276f3e1ec7bR278)) is https://github.com/ansible/ansible/pull/40105/files#diff-cb51550dcfdbd27a3f224be2de15e2e8R396. Either the config definition loading must happen first to work with config or class_only needs to be used so the definitions are loaded first.

s-hertel commented on test/units/vars/test_variable_manager.py:
```diff
@@ -21,8 +21,6 @@
 
 import os
 
-from collections import defaultdict
```
I could remove the changes to this file, it's irrelevant to the feature I've added. I just noticed it was masking the real fact_cache and so the tests weren't failing when I broke it.

bcoca replied:
we kind of mask it in variablemanager itself, but we should not in these tests

s-hertel commented on lib/ansible/vars/fact_cache.py:
```diff
@@ -21,9 +21,10 @@ class FactCache(MutableMapping):
 
-        self._plugin = cache_loader.get(C.CACHE_PLUGIN)
-        if not self._plugin:
+        _plugin = cache_loader.get(C.CACHE_PLUGIN, class_only=True)
```
I'm not sure how class_only was intended to be used and therefore what the right fix is. The alternative solution for this (and [this](https://github.com/ansible/ansible/pull/50446/files#diff-a8385538c61f1318541f0276f3e1ec7bR278)) is https://github.com/ansible/ansible/pull/40105/files#diff-cb51550dcfdbd27a3f224be2de15e2e8R396. Either the config definition loading must happen first to work with config or class_only needs to be used so the definitions are loaded first.

bcoca replied:
the use looks fine, I'm not sure why it is needed, did the previous code not raise the error correctly?

s-hertel replied:
It looks like you fixed this in 960388143f086bd229b3d5f779b7b9e2b78c8bd7 by moving the config loading first so I can remove the class_only stuff now. `get_option` and `set_options` rely on the definitions being loaded before creating an instance of the plugin. Those were not necessarily loaded first (causing this error https://github.com/ansible/ansible/blob/465df0ef8d27d7fd534f521335db11490ebd43bc/lib/ansible/config/manager.py#L462) and that's why this was necessary.

s-hertel committed:
Fix cache plugin _prefix defaults to be the same as when they were loaded from constants

s-hertel committed:
Remove outdated class_only necessity now that config definition loading happens in time

s-hertel committed:
Fix inventory_cache definition precedence

s-hertel committed:
Consistently name inventory caching options

s-hertel committed:
Add some dev guide documentation for cache plugins

Add user documentation for inventory caching

Add a porting guide for custom cache plugins since they should use self.set_options() and self.get_option(option_name) instead of ansible.constants

s-hertel committed:
Replace jsonify with AnsibleJSONEncoder

s-hertel committed:
Add integration test to check that fact caching works normally with cache plugins using ansible.constants and inventory caching provides a helpful error for non-compatible cache plugins

s-hertel committed:
Use valid alias

s-hertel committed:
Fix tests

s-hertel commented:
@acozine Can I get a review on this from a documentation perspective, please?

s-hertel commented:
Rebased for porting guide and doc fragment conflicts.

s-hertel committed:
Add deprecation docs

s-hertel committed:
Handle new exception raised instead

s-hertel commented on lib/ansible/plugins/cache/__init__.py:
```diff
@@ -102,7 +102,7 @@ def __init__(self, *args, **kwargs):
             self._cache_dir = self._get_cache_connection(self.get_option('_uri'))
             self._timeout = float(self.get_option('_timeout'))
-        except AnsibleError:
+        except KeyError:
```
changed in become plugins

s-hertel committed:
shorten line length

acozine commented on docs/docsite/rst/dev_guide/developing_inventory.rst:
```diff
@@ -189,9 +189,76 @@ To facilitate this there are a few of helper functions used in the example below
+inventory cache
+^^^^^^^^^^^^^^^
+
+If you used the Cacheable base class and have extended the inventory plugin documentation with the inventory_cache documentation fragment you have caching at your disposal.
```
Is this the only way to get caching? If so, I'd recommend stating this as a direction rather than "if/then" (because "if/then" suggests there's an alternative - "if not/then"):

"To enable caching, use the ``Cacheable`` base class and extend the inventory plugin documentation with the ``inventory_cache`` documentation fragment."

acozine commented on docs/docsite/rst/dev_guide/developing_inventory.rst:
```diff
@@ -189,9 +189,76 @@ To facilitate this there are a few of helper functions used in the example below
+
+If you used the Cacheable base class and have extended the inventory plugin documentation with the inventory_cache documentation fragment you have caching at your disposal.
+
+Now you can load the cache plugin specified by the user to read from and update the cache. If your inventory plugin uses YAML based configuration files and the ``_read_config_data`` method, the cache plugin is loaded within that method. Accordingly, you can skip this first step. Otherwise, for plugins that don't use ``_read_config_data``, the cache will need to be loaded with ``load_cache_plugin` ...
```
```suggestion
Next, load the cache plugin specified by the user to read from and update the cache. If your inventory plugin uses YAML based configuration files and the ``_read_config_data`` method, the cache plugin is loaded within that method. If your inventory plugin does not use ``_read_config_data``, you must load the cache explicitly with ``load_cache_plugin``.
```

acozine commented on docs/docsite/rst/dev_guide/developing_inventory.rst:
```diff
@@ -189,9 +189,76 @@ To facilitate this there are a few of helper functions used in the example below
+
+        self.load_cache_plugin()
+
+Before using the cache, retrieve a unique cache key using the ``get_cache_key`` method. This needs to be done by all inventory modules using the cache.
```
Information on why this needs to be done might help users here - for security? to enable two-way communication? Something like, "Before using the cache, you must retrieve a unique cache key . . . . method. This required step establishes a secure connection." (or whatever the actual reason).

acozine commented on docs/docsite/rst/dev_guide/developing_inventory.rst:
```diff
@@ -189,9 +189,76 @@ To facilitate this there are a few of helper functions used in the example below
+        self.load_cache_plugin()
+        cache_key = self.get_cache_key(path)
+
+The last parameter of ``parse`` is ``cache``. This value comes from the inventory manager and indicates whether the inventory is being refreshed (such as via ``--flush-cache`` or the meta task ``refresh_inventory``). Although the cache shouldn't be used to populate the inventory when being refreshed, the cache should be updated with the new inventory if the user has enabled caching. You can use ` ...
```
This paragraph contains good info, and a better introductory sentence would make it great. Unfortunately I'm not sure what the main topic is, so I'm not sure what to suggest. It might be: "Now that you've enabled caching and loaded the correct plugin, you can set up the flow of data between the cache and your inventory using the ``cache`` parameter of the ``parse`` method." Or it might be:  "Be careful with the ``cache`` parameter, which controls data flow into and out of the cache; you want to avoid overwriting inventory at the wrong times." Again, detailing why this information is important helps users (and me!) understand it.

s-hertel replied:
I think I made this better. The topic is the flow of data/logic.

acozine commented on docs/docsite/rst/dev_guide/developing_inventory.rst:
```diff
@@ -189,9 +189,76 @@ To facilitate this there are a few of helper functions used in the example below
+
+After the ``parse`` method is complete, the contents of ``self._cache`` is used to set the cache plugin if the contents of the cache have changed.
+
+You also have cache methods ``set_cache_plugin``, ``update_cache_if_changed``, and ``clear_cache`` available. If you need to set the cache plugin before ``parse`` completes for some reason, you can use ``self.set_cache_plugin()`` which will force the cache plugin to be set with the contents of ``self._cache``. Alternatively, ``self.update_cache_if_changed()`` can be used to only set the cache plu ...
```
Here I'd either fill in the "why" or make this a list, like this:
```
You have three other cache methods available:
- ``set_cache_plugin`` forces the cache plugin to be set with the contents of ``self._cache``
- ``update_cache_if_changed`` sets the cache plugin only if ``self._cache`` has been modified
- ``clear_cache`` deletes the keys in ``self._cache`` from your cache plugin
```

acozine reviewed (commented):
Lots of nice work here - thanks for adding all this documentation. This review makes some suggestions on the developing_inventory parts. I'll look at the plugin dev parts tomorrow.

acozine committed:
Update docs/docsite/rst/dev_guide/developing_inventory.rst

Co-Authored-By: s-hertel <shertel@redhat.com>

s-hertel committed:
Cleanup dev docs

acozine commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -144,6 +146,46 @@ the local time, returning the time delta in days, seconds and microseconds.
+
+Cache plugins store gathered facts and data retrieved by inventory plugins.
+
+Cache plugins should not be imported directly in the code base. Import cache plugins using the cache_loader so ``self.set_options()`` and ``self.get_option(<option_name>)`` can be used. Not using the cache_loader would restrict options to be only be available via ``ansible.constants``. This would also break its ability to be used by an inventory plugin.
```
```suggestion
Import cache plugins using the cache_loader so you can use ``self.set_options()`` and ``self.get_option(<option_name>)``. If you import a cache plugin directly in the code base, you can only access options via ``ansible.constants``, and you break the cache plugin's ability to be used by an inventory plugin.
```

acozine commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -144,6 +146,46 @@ the local time, returning the time delta in days, seconds and microseconds.
+            self._prefix = self.get_option('_prefix')
+            self._timeout = self.get_option('_timeout')
+
+Caches that use ``BaseCacheModule`` will need to implement the methods ``get``, ``contains``, ``keys``, ``set``, ``delete``, ``flush``, and ``copy``. The ``contains`` method should return a boolean that indicates if the key exists and has not expired. Unlike file-based caches, the ``get`` method does not raise a KeyError if the cache has expired.
```
```suggestion
If you use the ``BaseCacheModule``, you must implement the methods ``get``, ``contains``, ``keys``, ``set``, ``delete``, ``flush``, and ``copy``. The ``contains`` method should return a boolean that indicates if the key exists and has not expired. Unlike file-based caches, the ``get`` method does not raise a KeyError if the cache has expired.
```

acozine commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -144,6 +146,46 @@ the local time, returning the time delta in days, seconds and microseconds.
+
+Caches that use ``BaseCacheModule`` will need to implement the methods ``get``, ``contains``, ``keys``, ``set``, ``delete``, ``flush``, and ``copy``. The ``contains`` method should return a boolean that indicates if the key exists and has not expired. Unlike file-based caches, the ``get`` method does not raise a KeyError if the cache has expired.
+
+Caches that use ``BaseFileCacheModule`` must implement ``_load`` and ``_dump`` methods that will be called from the base class methods ``get`` and ``set``.
```
```suggestion
If you use the ``BaseFileCacheModule``, you must implement ``_load`` and ``_dump`` methods that will be called from the base class methods ``get`` and ``set``.
```

acozine commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -144,6 +146,46 @@ the local time, returning the time delta in days, seconds and microseconds.
+
+Caches that use ``BaseFileCacheModule`` must implement ``_load`` and ``_dump`` methods that will be called from the base class methods ``get`` and ``set``.
+
+If the cache plugin stores JSON, ``AnsibleJSONEncoder`` should be used in the ``_dump`` or ``set`` method  and ``AnsibleJSONDecoder`` should be used in the ``_load`` or ``get`` method.
```
```suggestion
If your cache plugin stores JSON, use ``AnsibleJSONEncoder`` in the ``_dump`` or ``set`` method  and ``AnsibleJSONDecoder`` in the ``_load`` or ``get`` method.
```

acozine commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -110,6 +110,31 @@ Now the output of ``ansible-inventory -i demo.aws_ec2.yml --graph``:
 
 If a host does not have the variables in the configuration above (i.e. ``tags.Name``, ``tags``, ``private_ip_address``), the host will not be added to groups other than those that the inventory plugin creates and the ``ansible_host`` host variable will not be modified.
 
+If an inventory plugin supports caching you can enable and set options for that per YAML configuration source or for multiple inventory sources using environment variables or Ansible configuration files. If caching has been enabled for an inventory plugin but no inventory-specific caching options were provided, fact caching options are used. Here is an example of enabling caching for an individua ...
```
```suggestion
If an inventory plugin supports caching, you can enable and set caching options for an individual YAML configuration source or for multiple inventory sources using environment variables or Ansible configuration files. If you enable caching for an inventory plugin without providing inventory-specific caching options, the inventory plugin will use fact-caching options. Here is an example of enabling caching for an individual YAML configuration file:
```

acozine commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -107,6 +107,19 @@ Deprecated
      removed in 2.12.  If you need the old behaviour switch to ``FactCache.first_order_merge()``
      instead.
 
+* Inventory plugins writers should know that supporting file-backed caching via self.cache is deprecated and will
```
```suggestion
* Supporting file-backed caching via self.cache is deprecated and will
```

acozine commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -107,6 +107,19 @@ Deprecated
      instead.
 
+* Inventory plugins writers should know that supporting file-backed caching via self.cache is deprecated and will
+  be removed in Ansible 2.12. Instead, self._cache should be used as a dictionary. For implementation details, see
```
```suggestion
  be removed in Ansible 2.12. If you maintain an inventory plugin, update it to use ``self._cache`` as a dictionary. For implementation details, see
```

acozine commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -107,6 +107,19 @@ Deprecated
 
+* Inventory plugins writers should know that supporting file-backed caching via self.cache is deprecated and will
+  be removed in Ansible 2.12. Instead, self._cache should be used as a dictionary. For implementation details, see
+  the developer guide on inventory plugins.
```
```suggestion
  the :ref:`developer guide on inventory plugins<inventory_plugin_caching>`.
```

acozine commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -107,6 +107,19 @@ Deprecated
+  be removed in Ansible 2.12. Instead, self._cache should be used as a dictionary. For implementation details, see
+  the developer guide on inventory plugins.
+
+* Importing cache plugins directly is deprecated and will be removed in Ansible 2.12. The plugin_loader should be
```
```suggestion
* Importing cache plugins directly is deprecated and will be removed in Ansible 2.12. Use the plugin_loader
```

acozine commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -107,6 +107,19 @@ Deprecated
+  the developer guide on inventory plugins.
+
+* Importing cache plugins directly is deprecated and will be removed in Ansible 2.12. The plugin_loader should be
+  used so direct options, environment variables, and other means of configuration can be reconciled using the config
```
```suggestion
  so direct options, environment variables, and other means of configuration can be reconciled using the config
```

acozine commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -232,6 +245,49 @@ Plugins
 
 * Play recap now counts ``ignored`` and ``rescued`` tasks as well as ``ok``, ``changed``, ``unreachable``, ``failed`` and ``skipped`` tasks, thanks to two additional stat counters in the ``default`` callback plugin. Tasks that fail and have ``ignore_errors: yes`` set are listed as ``ignored``. Tasks that fail and then execute a rescue section are listed as ``rescued``. Note that ``rescued`` tasks ...
 
+* Cache plugins can now be used for caching inventory. To make a new or custom cache plugin compatible with inventory
```
This section seems awfully long for the porting guide. I'd move anything that isn't already covered in the `developing_inventory` page to there, and link to it from here. That way folks will be able to find this info easily even when 2.8 starts to fade in the rear-view mirror.

s-hertel replied:
Good point!

acozine reviewed (commented):
Thanks @s-hertel for expanding the documentation on writing and maintaining inventory and cache plugins! I've added some more suggestions - let me know if they spark questions or raise concerns.

s-hertel committed:
Remove details for how to update a custom cache plugin from the porting guide

acozine committed:
Apply suggestions from code review

Co-Authored-By: s-hertel <shertel@redhat.com>

s-hertel commented:
Rebased for porting guide conflict.

s-hertel commented:
@FutureTecSystems It looks like you have some valuable pieces in lib/ansible/plugins/inventory/foreman.py in that PR. I think it might be out of the scope of this pull request. I've tried to limit my changes to only those necessary for getting cache plugins to be functional with inventory plugins, so optimizing foreman may be better as a follow up PR.

s-hertel committed:
Fix yaml block

s-hertel committed:
fix sanity test

s-hertel committed:
Add changelog

acozine commented on docs/docsite/rst/dev_guide/developing_inventory.rst:
```diff
@@ -189,9 +189,90 @@ To facilitate this there are a few of helper functions used in the example below
+        self.load_cache_plugin()
+        cache_key = self.get_cache_key(path)
+
+Now that you've enabled caching and loaded the correct plugin, you can set up the flow of data between the cache and your inventory using the `cache` parameter of the `parse` method. This value comes from the inventory manager and indicates whether the inventory is being refreshed (such as via ``--flush-cache`` or the meta task ``refresh_inventory``). Although the cache shouldn't be used to popul ...
```
```suggestion
Now that you've enabled caching, loaded the correct plugin, and retrieved a unique cache key, you can set up the flow of data between the cache and your inventory using the ``cache`` parameter of the ``parse`` method. This value comes from the inventory manager and indicates whether the inventory is being refreshed (such as via ``--flush-cache`` or the meta task ``refresh_inventory``). Although the cache shouldn't be used to populate the inventory when being refreshed, the cache should be updated with the new inventory if the user has enabled caching. You can use ``self._cache`` like a dictionary. The following pattern allows refreshing the inventory to work in conjunction with caching.
```

acozine commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -70,6 +70,8 @@ To define configurable options for your plugin, describe them in the ``DOCUMENTA
 
 To access the configuration settings in your plugin, use ``self.get_option(<option_name>)``. For most plugin types, the controller pre-populates the settings. If you need to populate settings explicitly, use a ``self.set_options()`` call.
 
+Cache plugins should always access options with ``self.get_option(<option_name>)`` to allow inventory plugins to use them.
```
```suggestion
```

acozine reviewed (commented):
Thanks @s-hertel for creating and updating the docs for this. Last docs changes suggested.

acozine committed:
Update docs/docsite/rst/dev_guide/developing_inventory.rst

Co-Authored-By: s-hertel <shertel@redhat.com>

acozine committed:
Update docs/docsite/rst/dev_guide/developing_plugins.rst

Co-Authored-By: s-hertel <shertel@redhat.com>

AlanCoding commented:
I'm hitting this message

```
[DEPRECATION WARNING]: InventoryModule should utilize self._cache as a dict 
instead of self.cache. To set the self._cache dictionary, use self._cache[key] 
= value instead of self.cache.set(key, value). To force update the underlying 
cache plugin with the contents of self._cache before parse() is complete, call 
self.set_cache_plugin and it will use the self._cache dictionary to update the 
cache plugin. This feature will be removed in version 2.12. Deprecation 
warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
```

with the openstack inventory plugin.

What does this mean? Is this automatically worthy of creating a new issue in your issue queue, since it looks like the message is directed at the maintainers of the inventory plugin itself, and the openstack inventory plugin is a part of Ansible.
